be/src/io/cache/sub_file_cache.cpp

```diff
@@ -121,48 +121,84 @@ Status SubFileCache::_generate_cache_reader(size_t offset, size_t req_size) {
     RETURN_NOT_OK_STATUS_WITH_WARN(
             io::global_local_filesystem()->exists(cache_done_file, &done_file_exist),
             fmt::format("Check local cache done file exist failed. {}", cache_done_file.native()));
+
+    std::promise<Status> download_st;
+    std::future<Status> future = download_st.get_future();
     if (!done_file_exist) {
-        bool cache_file_exist = false;
-        RETURN_NOT_OK_STATUS_WITH_WARN(
-                io::global_local_filesystem()->exists(cache_file, &cache_file_exist),
-                fmt::format("Check local cache file exist failed. {}", cache_file.native()));
-        if (cache_file_exist) {
-            RETURN_NOT_OK_STATUS_WITH_WARN(
-                    io::global_local_filesystem()->delete_file(cache_file),
-                    fmt::format("Check local cache file exist failed. {}", cache_file.native()));
+        ThreadPoolToken* thread_token =
+                ExecEnv::GetInstance()->get_serial_download_cache_thread_token();
+        if (thread_token != nullptr) {
+            auto st = thread_token->submit_func([this, &download_st, cache_done_file, cache_file,
+                                                 offset, req_size] {
+                auto func = [this, cache_done_file, cache_file, offset, req_size] {
+                    bool done_file_exist = false;
+                    // Judge again whether cache_done_file exists, it is possible that the cache
+                    // is downloaded while waiting in the thread pool
+                    RETURN_NOT_OK_STATUS_WITH_WARN(io::global_local_filesystem()->exists(
+                                                           cache_done_file, &done_file_exist),
+                                                   "Check local cache done file exist failed.");
+                    bool cache_file_exist = false;
+                    RETURN_NOT_OK_STATUS_WITH_WARN(
+                            io::global_local_filesystem()->exists(cache_file, &cache_file_exist),
+                            fmt::format("Check local cache file exist failed. {}",
+                                        cache_file.native()));
+                    if (!done_file_exist && cache_file_exist) {
```

Comment:
Don't return here. Download files after delete cache_file, and do the next work in this function.

Reply:
done